packages/astro/src/vite-plugin-astro/compile.ts

```diff
@@ -0,0 +1,147 @@
+import { fileURLToPath } from 'url';
+import { ESBuildTransformResult, transformWithEsbuild } from 'vite';
+import { AstroConfig } from '../@types/astro';
+import { cachedCompilation, CompileProps, CompileResult } from '../core/compile/index.js';
+import { LogOptions } from '../core/logger/core.js';
+import { getFileInfo } from '../vite-plugin-utils/index.js';
+
+interface CachedFullCompilation {
+	compileProps: CompileProps;
+	rawId: string;
+	logging: LogOptions;
+}
+
+interface FullCompileResult extends Omit<CompileResult, 'map'> {
+	map: ESBuildTransformResult['map'];
+}
+
+interface EnhanceCompilerErrorOptions {
+	err: Error;
+	id: string;
+	source: string;
+	config: AstroConfig;
+	logging: LogOptions;
+}
+
+const FRONTMATTER_PARSE_REGEXP = /^\-\-\-(.*)^\-\-\-/ms;
+
+export async function cachedFullCompilation({
+	compileProps,
+	rawId,
+	logging,
+}: CachedFullCompilation): Promise<FullCompileResult> {
+	let transformResult: CompileResult;
+	let esbuildResult: ESBuildTransformResult;
+
+	try {
+		transformResult = await cachedCompilation(compileProps);
+		// Compile all TypeScript to JavaScript.
+		// Also, catches invalid JS/TS in the compiled output before returning.
+		esbuildResult = await transformWithEsbuild(transformResult.code, rawId, {
```

Comment:
Swapping our use of `esbuild.transform` to `transformWithEsbuild` so we use Vite's esbuild instance, and get the benefits of respecting the tsconfig.

I'm planning to refactor this more for the JSX plugin to when I backport https://github.com/vitejs/vite/pull/11120 to Vite 3.

Reply:
Oh interesting, how was our esbuild not respecting it?

Reply:
Vite has [this code](https://github.com/vitejs/vite/blob/8d00673bdccdb9eec248f68bf790798f1d7e018e/packages/vite/src/node/plugins/esbuild.ts#L104) that passes the resolved tsconfig for that file during transform, as esbuild doesn't handle that by default. It should make the compiled output a bit more accurate, but in most case not noticable.